src/hv_anndata/accessors.py

```diff
@@ -0,0 +1,165 @@
+        return False
+
+    def isin(self, adata: AnnData) -> bool:
+        try:
+            self(adata)
+        except (IndexError, KeyError):
+            return False
+        return True
+
+
+@dataclass(frozen=True)
+class LayerAcc:
+    def __getitem__(self, k: str) -> LayerVecAcc:
+        return LayerVecAcc(k)
+
+
+@dataclass(frozen=True)
+class LayerVecAcc:
+    k: str
+
+    def __getitem__(self, i: IdxStr2D) -> AdPath:
+        def get(ad: AnnData) -> pd.api.extensions.ExtensionArray | NDArray[Any]:
+            return np.asarray(ad[i].layers[self.k])  # TODO: pandas
+
+        return AdPath(f"A.layers[{self.k!r}][{i[0]!r}, {i[1]!r}]", get)
+
+
+@dataclass(frozen=True)
+class MetaAcc:
+    ax: Literal["obs", "var"]
+
+    def __getitem__(self, k: str) -> AdPath:
+        def get(ad: AnnData) -> pd.api.extensions.ExtensionArray | NDArray[Any]:
+            if k == 'index':
+                return getattr(ad, self.ax).index
+            return getattr(ad, self.ax)[k]
+
+        return AdPath(f"A.{self.ax}[{k!r}]", get)
+
+
+@dataclass(frozen=True)
+class MultiAcc:
+    ax: Literal["obsm", "varm"]
+
+    def __getitem__(self, k: str) -> MultiVecAcc:
+        return MultiVecAcc(self.ax, k)
+
+
+@dataclass(frozen=True)
+class MultiVecAcc:
+    ax: Literal["obsm", "varm"]
+    k: str
+
+    def __getitem__(self, i: int) -> AdPath:
+        def get(ad: AnnData) -> pd.api.extensions.ExtensionArray | NDArray[Any]:
+            return getattr(ad, self.ax)[self.k][:, i]
+
+        return AdPath(f"A.{self.ax}[{self.k!r}][:, {i!r}]", get)
+
+
+@dataclass(frozen=True)
+class GraphAcc:
+    ax: Literal["obsp", "varp"]
+
+    def __getitem__(self, k: str) -> GraphVecAcc:
+        return GraphVecAcc(self.ax, k)
+
+
+@dataclass(frozen=True)
+class GraphVecAcc:
+    ax: Literal["obsp", "varp"]
+    k: str
+
+    def __getitem__(self, i: IdxInt2D) -> AdPath:
+        def get(ad: AnnData) -> pd.api.extensions.ExtensionArray | NDArray[Any]:
+            return getattr(ad, self.ax)[self.k][i]
+
+        return AdPath(f"A.{self.ax}[{self.k!r}][{i[0]!r}, {i[1]!r}]", get)
+
+
+@dataclass(frozen=True)
+class AdAc:
+    layers = LayerAcc()
+    obs = MetaAcc("obs")
+    var = MetaAcc("var")
+    obsm = MultiAcc("obsm")
+    varm = MultiAcc("varm")
+    obsp = GraphAcc("obsp")
+    varp = GraphAcc("varp")
+
+    def __getitem__(self, i: IdxStr2D) -> AdPath:
+        def get(ad: AnnData) -> pd.api.extensions.ExtensionArray | NDArray[Any]:
+            return np.asarray(ad[i].X)  # TODO: pandas
+
+        return AdPath(f"A[{i[0]!r}, {i[1]!r}]", get)
+
+    @classmethod
+    def resolve(cls, spec: str) -> AdPath:
+        inst = cls()
+        acc, *pieces = spec.split(".")
+        pieces = list(pieces)
+        if not hasattr(inst, acc):
+            return inst[tuple(int(acc), int(pieces[0]))]
+        current = getattr(inst, acc)
+        if isinstance(current, (GraphAcc, LayerAcc, MetaAcc, MultiAcc)):
+            current = current[pieces.pop(0)]
```

Comment:
what if we want to access `obsm["foo.bar"][:, 1]`? Either this needs to handle this unambiguously or we just don’t allow it and raise an explicit error.

I don‘t think raising an error that says nonsense like “`obsm` doesn’t have key `foo`” here is a good idea.